### ISSUE TITLE
Add saner ByteString instances

### ISSUE THREAD

**Issue by Fuuzetsu** [bos/aeson#187]
ByteString instances were removed because they allowed people to try and insert arbitrary binary data that JSON can't handle.

Could we not re-add ByteString instances that leveraged something like base64 encoding (which JSON can take I believe) to work around this limitation?

```
<mgsloan> Fuuzetsu: Best thing to do is to avoid putting bytestrings in json,
          but when it's unavoidable:
          https://gist.github.com/mgsloan/bfe4bf2250c1d656a2b1  [05:03]
```

**Comment by phadej** [comment]
FWIW we also have [`ByteString64`](http://docs.futurice.com/haskell-mega-repo/base64-bytestring-type-0.1.0.0/Data-ByteString-Base64-Type.html).

OTOH @basvandijk's could be interesting, but I'd rather do it only for `template-haskell` derivations first.

**Comment by winterland1989** [comment]
+1 on having this instance exported from aeson, a dependency on base64-bytestring is acceptable.

**Comment by bergmark** [comment]
I don't have an opinon on this. It looks like there are 4 votes for adding this newtype and one against. Another option is to add this instance to a blessed orphans package (#432). OTOH the dependency is very light. Sounds like the first step should be to add this newtype to base64-bytestring no matter how we choose to proceed.

**Comment by cliffordbeshers** [comment]
I ran into this issue yesterday.  Comments and observations.

@basvandijk writes "Ideally types tell us how values should be used on the Haskell side, not how they are encoded in some particular medium."  I would like to see a larger justification for this statement, because it runs against much that I hold to be true.

In any case, I do not think it applies in this instance.  I am wrapping a newtype around my Bytestring precisely because I want the encoding/decoding of my data to be fully encapsulated in the aeson process.  Otherwise, I have to add an extra step on either the client or the server side everywhere I use that type.  This would be particularly onerous because the uses are asymmetrical.  On the server side, compiling with ghc, this bytestring gets read from the file system.  On the client side, compiling with ghcjs, it gets converted to a blob.  I have used additional encoding/decoding functions as you suggest, in fact, I have those exact ones, and they are error prone, becuse they do not encode he semantics of the encoding.  They are tricky to maintain and difficult for othesr to understand.  It exists in our production system and it causes me dread every time I get near it.

I will give the newtype solution a try today to see if I run into anything ugly.

**Comment by Fuuzetsu** [comment]
AFAIK the "canonical" solution to this for last few years is `ByteString64` newtype though I would still like it to live inside aeson as `ByteString` instance as it's often a hassle when you find out late into existing codebase that you'd like JSON instances and you end up rewrapping throughout whole codebase: nothing major but inconvenient.

**Comment by phadej** [comment]
@Fuuzetsu IMHO preferring single encoding is wrong a choice. Someone might as well have data in base16 (hex) too. It's a bit like `Sum` & `Product` for `Monoid`.

Wrapping business can be solved independently of `aeson`. (I can imagine using `generics-sop` `Code`s + `Coercible` to be able to do oneliner conversion).

**Comment by Fuuzetsu** [comment]
if you want different encoding then you can use newtype, this issue is about having a default

**Comment by basvandijk** [comment]
While thinking about this again I would like to slightly adapt my previous opinion on the matter:

Usually when you have a string of binary data somewhere in your record it represents a particular thing, i.e.: an image, blob, hash, ASCII text, etc.. In those cases wouldn't it be better to wrap that string of bytes with a *specific* newtype telling your users what the bytes mean?

So I guess I'm advocating using a dedicated newtype for each type of bytestring instead of using a `ByteString` or `ByteString64` directly. For example, instead of having:

```haskell
data MyRecord = MyRecord
  { ...
  , blob :: ByteString
  , hash :: ByteString
  , ...
  }
```

You should have:

```haskell
data MyRecord = MyRecord
  { ...
  , blob :: Blob
  , hash :: Hash
  , ...
  }
```

where

```haskell
newtype Blob = Blob ByteString

instance ToJSON Blob where ...
instance FromJSON Blob where ...

newtype Hash = Hash ByteString

instance ToJSON Hash where ...
instance FromJSON Hash where ...

```

In the latter it wouldn't hurt using:

```haskell
newtype Blob = Blob ByteString64 deriving (ToJSON, FromJSON)
newtype Hash = Hash ByteString64 deriving (ToJSON, FromJSON)
```

to get the JSON instances for free.

What do you folks think about this? Would this be to much hassle in practise?

**Comment by Fuuzetsu** [comment]
it's no different than `blob :: ByteString64, hash :: ByteString64` except now you need an instance for every one of your wrappers (Blob, Hash, …) that you have to write rather than having 1 instance. So it seems strictly worse from the "amount of code" perspective. Of course knowing we can't swap Blob and Hash around by accident has its own merits but unsure if we can use that as an argument here…

**Comment by basvandijk** [comment]
Yes you do get an extra newtype wrapper W (```newtype W = W ByteString64 deriving (ToJSON, FromJSON)```) for every different type of bytestring which indeed causes more code.

BTW I'm not opposing providing a `ByteString64` newtype. Exporting it somewhere from `base64-bytestring` would be nice.

**Comment by bgamari** [comment]
For the record, @phadej has uploaded his `base64-bytestring-type` package to [Hackage](https://hackage.haskell.org/package/base64-bytestring-type).

**Comment by phadej** [comment]
@bgamari well spotted!

**Comment by hvr** [comment]
...or good stalking ;-)

**Comment by djleonskennedy** [comment]
@basvandijk Might i miss something but type says:
`encode :: ToJSON a => a -> ByteString` -- **ByteString** 
and then: 
`No instance for (ToJSON ByteString)`
How to deal with it without workarounds with generics, implement own instance?
Regards

**Comment by bgamari** [comment]
@djleonskennedy, can you rephrase the question? I'm not sure I follow.
